internal/cluster/cluster_service.go

```diff
@@ -154,28 +154,31 @@ func (s *Service) sendMemberListViewRequest(ctx context.Context, conn *Connectio
 }
 
 type AddrSet struct {
-	addrs map[string]pubcluster.Address
+	addrs       map[string]struct{}
+	orderedAddr []pubcluster.Address
 }
 
-func NewAddrSet() AddrSet {
-	return AddrSet{addrs: map[string]pubcluster.Address{}}
+func NewAddrSet() *AddrSet {
```

Comment:
Is there a reason to return a pointer?

Reply:
`AddrSet` contains a slice. It is updated with `AddAddr` method. If we do not use a pointer receiver, updates to the slice are not reflected to the instance.

Reply:
We should use a pointer receiver with `AddAddr` but we can still return non-pointer `AddrSet` from `NewAddrSet`, they are not related.

Reply:
I find it intuitive for constructors to return pointers to the original struct if that struct implements its method with a pointer receiver. It is harder to misuse on caller's side. Is there a reason not to return a pointer?

Reply:
In this case, the code is not in the hot path, and Go can probably optimize the `AddrSet` instance to be allocated on the stack but it's generally preferred to create non-pointer values if possible in order to reduce GC pressure. I am not sure how the user can misuse `AddrSet` when it's not a pointer, but converting it a pointer makes it easier to make mistakes:
```
func NewAddrSet() AddrSet {
    // ...
}
a1 := NewAddrSet()
for i := 0; i < 10; i++ {
	a1.AddAddr(fmt.Sprintf("%d.com", i))
}
b1 := a1
for i := 10; i < 100; i++ {
	b1.AddAddr(fmt.Sprintf("%d.com", i))
}
fmt.Println(len(a1.Addrs()))
```
Prints 10 as expected, but:
```
func NewAddrSet() *AddrSet {
    // ...
}
a1 := NewAddrSet()
for i := 0; i < 10; i++ {
	a1.AddAddr(fmt.Sprintf("%d.com", i))
}
b1 := a1
for i := 11; i < 100; i++ {
	b1.AddAddr(fmt.Sprintf("%d.com", i))
}
fmt.Println(len(a1.Addrs()))
```
prints 100

Reply:
I think implicitly, value receiver conveys the meaning "it is safe to copy me". For a type such AddrSet, I expect it to be statefult, meaning that it should recall the things I inserted even if I copy it.
I'm not so sure about re-assignment. I don't think it is a common usecase. On the other hand, passing custom types to  functions as parameters is. In that case, pointer receiver automatically behaves as expected.
I agree that performance shouldn't be part of the discussion for this case.
Please correct me if I'm wrong but I think there is no strong argument for it to be value receiver.

Reply:
After some discussion, we decided to refactor AddrSet type into a function to avoid any misuse in the future https://github.com/hazelcast/hazelcast-go-client/pull/885/commits/d951ef7360b3eff55bf4fcb3c3e0cb7b87942f93